tools/cell_census_builder/validate.py

```diff
@@ -271,39 +274,82 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
         with open_experiment(soma_path, eb) as exp:
             anndata_cell_filter = make_anndata_cell_filter(eb.anndata_cell_filter_spec)
             ad = anndata_cell_filter(unfiltered_ad, retain_X=True)
+            if ad.raw is None:
+                expected_ad_x, expected_ad_var = ad.X, ad.var
+            else:
+                expected_ad_x, expected_ad_var = ad.raw.X, ad.raw.var
 
-            soma_joinids: npt.NDArray[np.int64] = (
+            # get the joinids for the obs axis
+            obs_joinids = (
                 exp.obs.read(
                     column_names=["soma_joinid", "dataset_id"], value_filter=f"dataset_id == '{dataset.dataset_id}'"
                 )
                 .concat()
                 .to_pandas()
-                .soma_joinid.to_numpy()
             )
 
-            raw_nnz = 0
-            if len(soma_joinids) > 0:
-                assert soma.SparseNDArray.exists(exp.ms["RNA"].X["raw"].uri)
-
-                def count_elements(arr: soma.SparseNDArray, join_ids: npt.NDArray[np.int64]) -> int:
-                    return sum(t.non_zero_length for t in arr.read((join_ids, slice(None))).coos())
-
-                raw_nnz = count_elements(exp.ms["RNA"].X["raw"], soma_joinids)
+            assert ad.n_obs == len(obs_joinids)
+            if len(obs_joinids) == 0:
+                continue
 
-            def count_nonzero(arr: Union[sparse.spmatrix, npt.NDArray[Any]]) -> int:
-                """Return _actual_ non-zero count, NOT the stored value count."""
-                if isinstance(arr, (sparse.spmatrix, sparse.coo_array, sparse.csr_array, sparse.csc_array)):
-                    return np.count_nonzero(arr.data)
-                return np.count_nonzero(arr)
+            X_raw = (
+                exp.ms[MEASUREMENT_RNA_NAME]
```

Comment:
super small item:  you might want to change all of the hard-coded `"RNA"` strings to use the global `MEASUREMENT_RNA_NAME`, so the code is consistent.

I apologize in advance for my inconsistent coding!